File: projects/06/app/src/test/java/io/github/ironking24/nand2tetris/HackAssembler/CodeModuleTest.java
package io.github.ironking24.nand2tetris.HackAssembler;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrowsExactly;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import org.apache.commons.lang3.RandomStringUtils;




@DisplayName("Code Module Test")
public class CodeModuleTest {
    @Test
    @DisplayName("test comp")
    void compTest(){
        String randomString = RandomStringUtils.random(1000);
        
        assertAll("Computation instruction evaluation",
        ()->assertEquals("0101010", CodeModule.comp("0")),
        ()->assertEquals("0111111", CodeModule.comp("1")),
        ()->assertEquals("0111010", CodeModule.comp("-1")),
        ()->assertEquals("0001100", CodeModule.comp("D")),
        ()->assertEquals("0110000", CodeModule.comp("A")),
        ()->assertEquals("1110000", CodeModule.comp("M")),
        ()->assertEquals("0001101", CodeModule.comp("!D")),
        ()->assertEquals("0110001", CodeModule.comp("!A")),
        ()->assertEquals("1110001", CodeModule.comp("!M")),
        ()->assertEquals("0001111", CodeModule.comp("-D")),
        ()->assertEquals("0110011", CodeModule.comp("-A")),
        ()->assertEquals("1110011", CodeModule.comp("-M")),
        ()->assertEquals("0011111", CodeModule.comp("D+1")),
        ()->assertEquals("0110111", CodeModule.comp("A+1")),
        ()->assertEquals("1110111", CodeModule.comp("M+1")),
        ()->assertEquals("0001110", CodeModule.comp("D-1")),
        ()->assertEquals("0110010", CodeModule.comp("A-1")),
        ()->assertEquals("1110010", CodeModule.comp("M-1")),
        ()->assertEquals("0000010", CodeModule.comp("D+A")),
        ()->assertEquals("1000010", CodeModule.comp("D+M")),
        ()->assertEquals("0010011", CodeModule.comp("D-A")),
        ()->assertEquals("1010011", CodeModule.comp("D-M")),
        ()->assertEquals("0000111", CodeModule.comp("A-D")),
        ()->assertEquals("1000111", CodeModule.comp("M-D")),
        ()->assertEquals("0000000", CodeModule.comp("D&A")),
        ()->assertEquals("1000000", CodeModule.comp("D&M")),
        ()->assertEquals("0010101", CodeModule.comp("D|A")),
        ()->assertEquals("1010101", CodeModule.comp("D|M")),
        ()->assertThrowsExactly(IllegalArgumentException.class,
        ()->{CodeModule.comp(randomString);}
            ,"Illegal computation operation "+ randomString));
    }

    @Test
    @DisplayName("Test dest")
    void destTest(){
        String randomString = RandomStringUtils.random(1000);

        assertAll("Destination instruction evaluation",
        ()->assertEquals("000", CodeModule.dest("")),
        ()->assertEquals("001", CodeModule.dest("M")),
        ()->assertEquals("010", CodeModule.dest("D")),
        ()->assertEquals("011", CodeModule.dest("DM")),
        ()->assertEquals("100", CodeModule.dest("A")),
        ()->assertEquals("101", CodeModule.dest("AM")),
        ()->assertEquals("110", CodeModule.dest("AD")),
        ()->assertEquals("111", CodeModule.dest("ADM")),
        ()->assertThrowsExactly(IllegalArgumentException.class,
        ()->{CodeModule.comp(randomString);}
            ,"There is no destination as "+ randomString));
    }

    @Test
    @DisplayName("Test Jump")
    void jumpTest(){
        String randomString = RandomStringUtils.random(1000);

        assertAll(
            "Jump instruction evaluation",
            ()->assertEquals("000", CodeModule.jump("")),
            ()->assertEquals("001", CodeModule.jump("JGT")),
            ()->assertEquals("010", CodeModule.jump("JEQ")),
            ()->assertEquals("011", CodeModule.jump("JGE")),
            ()->assertEquals("100", CodeModule.jump("JLT")),
            ()->assertEquals("101", CodeModule.jump("JNE")),
            ()->assertEquals("110", CodeModule.jump("JLE")),
            ()->assertEquals("111", CodeModule.jump("JMP")),
            ()->assertThrowsExactly(IllegalArgumentException.class,
            ()->{CodeModule.comp(randomString);}
                ,"Illegal jump operation: "+ randomString));
    }
}

File: projects/06/test/src/CodeModuleTest.cpp
#pragma once
#include "gtest/gtest.h"
#include "..\..\src\CodeModule.h"

TEST(CodeModuleTest, DestinationTest)
{
	HackAssembler::CodeModule code_module = HackAssembler::CodeModule();

	ASSERT_EQ(code_module.dest("null"), "000");
	ASSERT_EQ(code_module.dest("M"), "001");
	ASSERT_EQ(code_module.dest("D"), "010");
	ASSERT_EQ(code_module.dest("DM"), "011");
	ASSERT_EQ(code_module.dest("A"), "100");
	ASSERT_EQ(code_module.dest("AM"), "101");
	ASSERT_EQ(code_module.dest("AD"), "110");
	ASSERT_EQ(code_module.dest("ADM"), "111");

	ASSERT_THROW(code_module.dest("random"), std::runtime_error);
}

TEST(CodeModuleTest, ComputationTest)
{
	HackAssembler::CodeModule code_module = HackAssembler::CodeModule();

	ASSERT_EQ(code_module.comp("0"), "0101010");
	ASSERT_EQ(code_module.comp("1"), "0111111");
	ASSERT_EQ(code_module.comp("-1"), "0111010");
	ASSERT_EQ(code_module.comp("D"), "0001100");
	ASSERT_EQ(code_module.comp("A"), "0110000");
	ASSERT_EQ(code_module.comp("!D"), "0001101");
	ASSERT_EQ(code_module.comp("!A"), "0110001");
	ASSERT_EQ(code_module.comp("-D"), "0001111");
	ASSERT_EQ(code_module.comp("-A"), "0110011");
	ASSERT_EQ(code_module.comp("D+1"), "0011111");
	ASSERT_EQ(code_module.comp("A+1"), "0110111");
	ASSERT_EQ(code_module.comp("D-1"), "0001110");
	ASSERT_EQ(code_module.comp("A-1"), "0110010");
	ASSERT_EQ(code_module.comp("D+A"), "0000010");
	ASSERT_EQ(code_module.comp("D-A"), "0010011");
	ASSERT_EQ(code_module.comp("A-D"), "0000111");
	ASSERT_EQ(code_module.comp("D&A"), "0000000");
	ASSERT_EQ(code_module.comp("D|A"), "0010101");

	ASSERT_EQ(code_module.comp("M"), "1110000");
	ASSERT_EQ(code_module.comp("!M"), "1110001");
	ASSERT_EQ(code_module.comp("-M"), "1110011");
	ASSERT_EQ(code_module.comp("M+1"), "1110111");
	ASSERT_EQ(code_module.comp("M-1"), "1110010");
	ASSERT_EQ(code_module.comp("D+M"), "1000010");
	ASSERT_EQ(code_module.comp("D-M"), "1010011");
	ASSERT_EQ(code_module.comp("M-D"), "1000111");
	ASSERT_EQ(code_module.comp("D&M"), "1000000");
	ASSERT_EQ(code_module.comp("D|M"), "1010101");

	ASSERT_THROW(code_module.comp("random"), std::runtime_error);
}

TEST(CodeModuleTest, JumpTest)
{
	HackAssembler::CodeModule code_module = HackAssembler::CodeModule();

	ASSERT_EQ(code_module.jump("null"), "000");
	ASSERT_EQ(code_module.jump("JGT"), "001");
	ASSERT_EQ(code_module.jump("JEQ"), "010");
	ASSERT_EQ(code_module.jump("JGE"), "011");
	ASSERT_EQ(code_module.jump("JLT"), "100");
	ASSERT_EQ(code_module.jump("JNE"), "101");
	ASSERT_EQ(code_module.jump("JLE"), "110");
	ASSERT_EQ(code_module.jump("JMP"), "111");

	ASSERT_THROW(code_module.jump("random"), std::runtime_error);
}